### PR TITLE
BUGFIX: SD-532 negative heatmap cells height error

### PR DESCRIPTION
**Pull request by lanhq**
[Check PR owner responsibilities](https://confluence.intgdc.com/display/Development/Code-reviews#Code-reviews-Ownerresponsibilities)

Supported PR commands:

Command | Description
--- | ---
`extended test - examples` | Live examples tests
`extended test - storybook` | Storybook screenshot tests

See more [options](https://confluence.intgdc.com/display/kbhr/How+to+work+with+git+and+Github#HowtoworkwithgitandGithub-Extendedchecks).

# PR Checklist

- [x] Verify code changes ([Checklist](https://confluence.intgdc.com/display/Development/Code-reviews+checklist), [Best practices](https://confluence.intgdc.com/display/Development/Code-reviews+best+practices))
- [x] [Verify pull-request formalities](https://confluence.intgdc.com/display/Development/Code-reviews)
- [x] Change was tested by using [gdc-dev-release](https://confluence.intgdc.com/display/~tomas.vojtasek/Private+NPM) in [gdc-analytical-designer](https://github.com/gooddata/gdc-analytical-designer) and [gdc-dashboards](https://github.com/gooddata/gdc-dashboards) (if applicable)
- [x] Migration guide (for major changes) is mentioned in [CHANGELOG.md](../blob/master/CHANGELOG.md).
- [x] Successful `extended test - examples`
- [x] Successful `extended test - storybook`
- [x] Checked yarn.lock consistency (no dep. duplicities especially Goodstrap)

# Related PRs

- gdc-analytical-designer: https://github.com/gooddata/gdc-analytical-designer/pull/2348
- gdc-dashboards: https://github.com/gooddata/gdc-dashboards/pull/2120

# Related Jira tasks
- [SD-532](https://jira.intgdc.com/browse/SD-532)

### CODE REVIEW

**Comment by lanhq** [comment]
extended test - storybook

**Comment by yenkins** [comment]
Link to Storybook test result:

https://checklist.intgdc.com/job/client-libs/job/gooddata-react-components-storybook-zuul-docker/2296/artifact/test-storybook/html-report/index.html

**Comment by gdgate** [comment]
Build succeeded (check-extended pipeline).

- https://checklist.intgdc.com/job/client-libs/job/gooddata-react-components-pull-request-dispatcher/1663/ : SUCCESS in 14m 05s
- https://checklist.intgdc.com/job/compliance/job/gooddata-react-components-opensource-repository-check/5433/ : SUCCESS in 10s

**Comment by gdgate** [comment]
Build failed (check pipeline).

- https://checklist.intgdc.com/job/client-libs/job/gooddata-react-components-unit-tests-zuul-docker-el7/4268/ : FAILURE in 15m 41s
- https://checklist.intgdc.com/job/compliance/job/gooddata-react-components-opensource-repository-check/5432/ : SUCCESS in 13s

**Comment by lanhq** [comment]
extended test - examples

**Comment by yenkins** [comment]
:arrow_right: Started Live Examples deploy to:
https://gdc-live-examples-1074.herokuapp.com

It will be destroyed on merge/close. See [ci-infra](https://github.com/gooddata/ci-infra/blob/master/jenkins/jobs/templates/orchestrators/client-pr-dispatcher.groovy) and [gdc-ci.yaml](https://github.com/gooddata/gooddata-react-components/blob/master/gdc-ci.yaml).

**Comment by gdgate** [comment]
Build succeeded (check pipeline).

- https://checklist.intgdc.com/job/client-libs/job/gooddata-react-components-unit-tests-zuul-docker-el7/4270/ : SUCCESS in 16m 26s
- https://checklist.intgdc.com/job/compliance/job/gooddata-react-components-opensource-repository-check/5436/ : SUCCESS in 31s

**Comment by gdgate** [comment]
Build succeeded (check-extended pipeline).

- https://checklist.intgdc.com/job/client-libs/job/gooddata-react-components-pull-request-dispatcher/1665/ : SUCCESS in 43m 21s
- https://checklist.intgdc.com/job/compliance/job/gooddata-react-components-opensource-repository-check/5437/ : SUCCESS in 12s

File: src/components/visualizations/chart/highcharts/heatmapConfiguration.ts [hunk 1]
```diff
@@ -1,7 +1,55 @@
...
+}
+
+function resetPointPaddingForTooSmallHeatmapCells(series: IHeatmapSeries) {
+    const xAxis: IHighchartsAxisExtend = series.xAxis;
+    const yAxis: IHighchartsAxisExtend = series.yAxis;
+    const options: Highcharts.SeriesOptionsType = series.options;
+    const seriesPointPadding: number = options.pointPadding || 0;
+    const pointPlacement: number = series.pointPlacementToXValue();
+
+    const between = (x: number, a: number, b: number) => {
```
Comment (naibinh) [inline src/components/visualizations/chart/highcharts/heatmapConfiguration.ts]: -> `(x: number, a: number, b: number): number => {...}`

**Comment by gdgate** [comment]
Build failed (check pipeline).

- gooddata-react-components-unit-tests-zuul-docker-el7 : SKIPPED
- https://checklist.intgdc.com/job/compliance/job/gooddata-react-components-opensource-repository-check/5462/ : Tests on opensource repositories should be triggered by users in `gooddata` organization in 3s

**Comment by gdgate** [comment]
Build failed (check pipeline).

- gooddata-react-components-unit-tests-zuul-docker-el7 : SKIPPED
- https://checklist.intgdc.com/job/compliance/job/gooddata-react-components-opensource-repository-check/5464/ : Tests on opensource repositories should be triggered by users in `gooddata` organization in 1s

**Comment by gdgate** [comment]
Build failed (check pipeline).

- gooddata-react-components-unit-tests-zuul-docker-el7 : SKIPPED
- https://checklist.intgdc.com/job/compliance/job/gooddata-react-components-opensource-repository-check/5465/ : Tests on opensource repositories should be triggered by users in `gooddata` organization in 2s

**Comment by naibinh** [comment]
LGTM

**Comment by lanhq** [comment]
ok to test

**Comment by gdgate** [comment]
Build failed (check pipeline).

- gooddata-react-components-unit-tests-zuul-docker-el7 : SKIPPED
- https://checklist.intgdc.com/job/compliance/job/gooddata-react-components-opensource-repository-check/5466/ : Tests on opensource repositories should be triggered by users in `gooddata` organization in 3s

**Comment by lanhq** [comment]
ok to test

**Comment by gdgate** [comment]
Build failed (check pipeline).

- https://checklist.intgdc.com/job/client-libs/job/gooddata-react-components-unit-tests-zuul-docker-el7/4281/ : FAILURE in 9m 00s
- https://checklist.intgdc.com/job/compliance/job/gooddata-react-components-opensource-repository-check/5467/ : SUCCESS in 13s

**Comment by gdgate** [comment]
Build succeeded (check pipeline).

- https://checklist.intgdc.com/job/client-libs/job/gooddata-react-components-unit-tests-zuul-docker-el7/4283/ : SUCCESS in 10m 39s
- https://checklist.intgdc.com/job/compliance/job/gooddata-react-components-opensource-repository-check/5470/ : SUCCESS in 6s

File: src/components/visualizations/chart/highcharts/heatmapConfiguration.ts [hunk 2]
```diff
@@ -1,7 +1,57 @@
 // (C) 2007-2019 GoodData Corporation
 import cloneDeep = require("lodash/cloneDeep");
 import last = require("lodash/last");
+import { IHighchartsAxisExtend } from "../../../../interfaces/HighchartsExtend";
 
+interface IHeatmapSeries extends Highcharts.Series {
+    pointPlacementToXValue?(): number;
+}
+
+function between(x: number, a: number, b: number): number {
```
Comment (ivan-nejezchleb) [inline src/components/visualizations/chart/highcharts/heatmapConfiguration.ts]: I would prefer better naming of limits. Something like lowerLimit, higherLimit

File: src/components/visualizations/chart/highcharts/heatmapConfiguration.ts [hunk 3]
```diff
@@ -1,7 +1,57 @@
...
+            2 * yAxis.len,
+        );
+        const y2 = between(
+            Math.round(yAxis.translate(point.y + yPad, 0, 1, 0, 1)),
+            -yAxis.len,
+            2 * yAxis.len,
+        );
+        const pointPadding: number = point.pointPadding || seriesPointPadding;
+
+        if (Math.abs(x2 - x1) < pointPadding * 2 || Math.abs(y2 - y1) < pointPadding * 2) {
```
Comment (ivan-nejezchleb) [inline src/components/visualizations/chart/highcharts/heatmapConfiguration.ts]: If I am right, this is only part changed in comparison with original HCH calculation. Mark it by some more explicit comment

File: src/components/visualizations/chart/highcharts/heatmapConfiguration.ts [hunk 4]
```diff
@@ -1,7 +1,57 @@
...
 
+interface IHeatmapSeries extends Highcharts.Series {
+    pointPlacementToXValue?(): number;
+}
+
+function between(x: number, a: number, b: number): number {
+    return Math.min(Math.max(a, x), b);
+}
+
+function resetPointPaddingForTooSmallHeatmapCells(series: IHeatmapSeries) {
```
Comment (ivan-nejezchleb) [inline src/components/visualizations/chart/highcharts/heatmapConfiguration.ts]: are we able to test this function somehow, with series mock?

**Comment by gdgate** [comment]
Build failed (check pipeline).

- https://checklist.intgdc.com/job/client-libs/job/gooddata-react-components-unit-tests-zuul-docker-el7/4398/ : FAILURE in 3m 27s
- https://checklist.intgdc.com/job/compliance/job/gooddata-react-components-opensource-repository-check/5637/ : SUCCESS in 11s

**Comment by gdgate** [comment]
Build failed (check pipeline).

- https://checklist.intgdc.com/job/client-libs/job/gooddata-react-components-unit-tests-zuul-docker-el7/4400/ : FAILURE in 1m 20s
- https://checklist.intgdc.com/job/compliance/job/gooddata-react-components-opensource-repository-check/5638/ : SUCCESS in 10s

**Comment by gdgate** [comment]
Build succeeded (check pipeline).

- https://checklist.intgdc.com/job/client-libs/job/gooddata-react-components-unit-tests-zuul-docker-el7/4401/ : SUCCESS in 5m 39s
- https://checklist.intgdc.com/job/compliance/job/gooddata-react-components-opensource-repository-check/5639/ : SUCCESS in 8s

File: src/components/visualizations/chart/highcharts/heatmapConfiguration.ts (hunk 3 above)
Comment (lanhq) [reply src/components/visualizations/chart/highcharts/heatmapConfiguration.ts]: comments is added.

**Comment by gdgate** [comment]
Build succeeded (check pipeline).

- https://checklist.intgdc.com/job/client-libs/job/gooddata-react-components-unit-tests-zuul-docker-el7/4405/ : SUCCESS in 11m 51s
- https://checklist.intgdc.com/job/compliance/job/gooddata-react-components-opensource-repository-check/5643/ : SUCCESS in 13s

**Comment by lanhq** [comment]
extended test - storybook

**Comment by gdgate** [comment]
Build in pipeline check-extended aborted.

**Comment by lanhq** [comment]
extended test - storybook

**Comment by gdgate** [comment]
Build succeeded (check pipeline).

- https://checklist.intgdc.com/job/client-libs/job/gooddata-react-components-unit-tests-zuul-docker-el7/4410/ : SUCCESS in 12m 53s
- https://checklist.intgdc.com/job/compliance/job/gooddata-react-components-opensource-repository-check/5647/ : SUCCESS in 9s

**Comment by yenkins** [comment]
Link to Storybook test result:

https://checklist.intgdc.com/job/client-libs/job/gooddata-react-components-storybook-zuul-docker/2368/artifact/test-storybook/html-report/index.html

**Comment by gdgate** [comment]
Build succeeded (check-extended pipeline).

- https://checklist.intgdc.com/job/client-libs/job/gooddata-react-components-pull-request-dispatcher/1755/ : SUCCESS in 11m 13s
- https://checklist.intgdc.com/job/compliance/job/gooddata-react-components-opensource-repository-check/5648/ : SUCCESS in 8s

**Comment by gdgate** [comment]
Build succeeded (gate pipeline).

- https://checklist.intgdc.com/job/client-libs/job/gooddata-react-components-storybook-zuul-docker/2374/ : SUCCESS in 13m 40s
- https://checklist.intgdc.com/job/client-libs/job/gooddata-react-components-unit-tests-zuul-docker-el7/4418/ : SUCCESS in 9m 44s
- https://checklist.intgdc.com/job/compliance/job/gooddata-react-components-merge-compliance-check/1119/ : SUCCESS in 26s
- https://checklist.intgdc.com/job/client-libs/job/gooddata-react-components-examples-zuul-docker/1387/ : SUCCESS in 24m 34s

**Pull request merged**